File: Data Structures & Algorithms/binary-tree-diameter/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(r):
            nonlocal res 
            if not r:
                return 0

            left_count = dfs(r.left)
            right_count = dfs(r.right)

            res = max(res, left_count + right_count)

            return 1 + max(left_count, right_count)

        dfs(root)

        return res

File: Data Structures & Algorithms/depth-of-binary-tree/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int MaxDepth(TreeNode root) {
        var stack = new Stack<(TreeNode, int)>();
        var maxCount = 0;

        stack.Push((root, 1));

        while (stack.Count > 0) {
            (var node, var count) = stack.Pop();

            if (node != null) {
                maxCount = Math.Max(maxCount, count);

                if (node.left != null) stack.Push((node.left, count + 1));
                if (node.right != null) stack.Push((node.right, count + 1));
            }
        }

        return maxCount;

        // if (root == null) return 0;

        // var left = MaxDepth(root.left) + 1;
        // var right = MaxDepth(root.right) + 1;

        // return Math.Max(left, right);
    }
}
